lib/Conversion/TorchToLinalg/TorchToLinalg.cpp

```diff
@@ -3435,6 +3435,132 @@ class ConvertAtenBroadcastToOp : public OpConversionPattern<AtenBroadcastToOp> {
+  auto tensorType = tensor.getType().cast<RankedTensorType>();
+  if (tensorType.getRank() == 0) {
+    Value one = b.create<arith::ConstantIndexOp>(loc, 1);
+    Value scalarVal = b.create<tensor::ExtractOp>(loc, tensor);
+    return createInitTensor(b, loc, {one}, tensorType.getElementType(),
+                            scalarVal);
+  }
+  if (tensorType.getRank() == 1)
+    return tensor;
+  SmallVector<ReassociationIndices> reassociation(1);
+  for (auto i : llvm::seq<int64_t>(0, tensorType.getRank()))
+    reassociation[0].push_back(i);
+  return b.create<linalg::TensorCollapseShapeOp>(loc, tensor, reassociation);
+}
+
+namespace {
+class ConvertAtenIndexPutOp : public OpConversionPattern<AtenIndexPutOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(AtenIndexPutOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
+      return failure();
+
+    Location loc = op.getLoc();
+    Value input = adaptor.self();
+    Value indices = op.indices();
+    Value elemVal = adaptor.values();
+    auto inputType = input.getType().cast<RankedTensorType>();
+    unsigned inputRank = inputType.getRank();
+
+    SmallVector<Value> indicesTuple;
+    if (!getListConstructElements(indices, indicesTuple)) {
+      return rewriter.notifyMatchFailure(
+          op, "unimplemented: the indices are not present.");
+    }
+    unsigned noOfIndices = indicesTuple.size();
+    SmallVector<Value> indicesVal =
+        getTypeConvertedValues(rewriter, loc, getTypeConverter(), indicesTuple);
+
+    // TODO: Support broadcasting along dim.
+    if (inputRank != noOfIndices)
+      return rewriter.notifyMatchFailure(
+          op, "invalid shape: Broadcasting value along dim is not supported");
+
+    // TODO: More comprehensive checks:-
+    //   1. Since broadcasting is not implemented the number of elements in each
+    //   indexes as well as the output tensor should be same.
+    //
+    //   2. Since broadcasting is not supported and no. of elements should be
+    //   same hence the shape of tensors should be equivalent, for example:
+    //   [x,y] and [x,y,1] or [x,y,1,1] is considered equivalent but [y,x] and
+    //   [x,y] are not considered equivalent.
+    //
+    //  3. There should be an assert that the index_values lies in the bounds of
+    //  the input tensor.
+
+    // The value tensor along with all the indices are collapsed to 1d tensor.
+    SmallVector<Value> collapsedIndices;
+    for (Value val : indicesVal)
+      collapsedIndices.push_back(collapseTo1dTensor(rewriter, loc, val));
+
+    elemVal = collapseTo1dTensor(rewriter, loc, elemVal);
+
+    AffineExpr zeroIdxExpr = rewriter.getAffineDimExpr(0);
+    // `indicesExpr` points to the mapping {d0,d1...} -> {d0} i.e., the
+    // outermost dimension.
+    SmallVector<AffineExpr, 4> indicesExpr{zeroIdxExpr};
```

Comment:
nit: no need to specify size of `SmallVector`. The same goes for all the other `SmallVector`s below

> In the absence of a well-motivated choice for the number of inlined elements N, it is recommended to use SmallVector<T> (that is, omitting the N).

https://www.llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h

Reply:
```../external/llvm-project/llvm/../mlir/include/mlir/IR/AffineMap.h:90:3: note: candidate function not viable: no known conversion from 'SmallVector<SmallVector<mlir::AffineExpr> >' to 'ArrayRef<SmallVector<mlir::AffineExpr, 4> >' for 1st argument```

Reply:
The function argument is expecting SmallVector<mlir::AffineExpr,4> instead of SmallVectorImpl.

Reply:
If you change the type of `indexExpr` here

https://github.com/llvm/torch-mlir/blob/b51330fb5f87241057a3e3356b028b1ebcd601c8/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L3813

to `SmallVector<ArrayRef<AffineExpr>>` it should work.

Conversion from `SmallVector` to `ArrayRef` happens automatically, however it does not happen recursively, so after this vector was converted, the result was an `ArrayRef<SmallVector<AffineExpr>>` rather than an `ArrayRef<ArrayRef<AffineExpr>>`, as expected by the AffineMap` function.

Reply:
@ramiro050 Thanks for the fix and explanation!